homework/zadanie4_five_philosophers.cpp

```diff
@@ -72,37 +56,62 @@ class CPhilosoph{
         }
     }
 };
+ 
+class CTable{
 
+    int numbers_of_forks_;
+    std::vector<std::thread> thread_pool_;
+    std::vector<std::string> names_of_philosophers_;
+    std::vector<CPhilosoph*> philo_pointers_;
+    std::vector<CFork*> forks_pool_;
 
-int main(int argc, char* argv[]){
-
-    const int numbers_of_philosophers = 3;
-    CTable table(numbers_of_philosophers);
-    std::vector<std::thread> thread_pool(numbers_of_philosophers);
-    std::vector<std::string> names_of_philosophers(numbers_of_philosophers);
-    std::vector<CPhilosoph*> philo_pointers;
-
-    for (int i=0; i< numbers_of_philosophers; i++){
-        names_of_philosophers[i] = "Philo" + std::to_string(i);
+public:
+    CTable(int number_of_forks) : numbers_of_forks_(number_of_forks){ 
+        for (int i=0; i < numbers_of_forks_; i++){
+            forks_pool_.push_back(new CFork());
+            names_of_philosophers_.push_back("Philo" + std::to_string(i));
+        }
     }
 
-    for (size_t i=0; i < numbers_of_philosophers; i++){
-        if (i == 0){
-            philo_pointers.push_back(new CPhilosoph(table, names_of_philosophers[i], table.forks_pool_[i], table.forks_pool_[numbers_of_philosophers-1]));
-            thread_pool[i] = std::thread(&CPhilosoph::party, philo_pointers[i]);
-        }
-        else{
-            philo_pointers.push_back(new CPhilosoph(table, names_of_philosophers[i], table.forks_pool_[i-1], table.forks_pool_[i]));
-            thread_pool[i] = std::thread(&CPhilosoph::party, philo_pointers[i]);
+    ~CTable(){
+        for (int i = 0; i < numbers_of_forks_; i++){
+            delete forks_pool_[i];
+            delete philo_pointers_[i];
         }
     }
-    for (auto&& th : thread_pool){
-        th.join();
+
+    void Start_party(){
+        for (size_t i=0; i < numbers_of_forks_; i++){
+            if (i == 0){
+                philo_pointers_.push_back(new CPhilosoph(names_of_philosophers_[i], forks_pool_[i], forks_pool_[numbers_of_forks_-1]));
+                thread_pool_.push_back(std::thread(&CPhilosoph::party, philo_pointers_[i]));
+            }
+            else{
+                philo_pointers_.push_back(new CPhilosoph(names_of_philosophers_[i], forks_pool_[i-1], forks_pool_[i]));
+                thread_pool_.push_back(std::thread(&CPhilosoph::party, philo_pointers_[i]));
+            }
+        }
     }
 
-    for (int i=0; i < numbers_of_philosophers; i++){
-        delete philo_pointers[i];
+    void Stop_party(){
+        
+        for (auto&& th : thread_pool_){
+            th.join();
+        }
+        
```

Comment:
to destruktor chyba powinien robić, wtedy mamy RAII. Bo tak jakby wyjątek poleciał to nie pozłącza Ci wątków i std::terminate może pojść.

Reply:
przeniesiono do destruktora,